<center>
  <h1>Reading Response | de Bruijn Graphs for Genome Assembly</h1>
</center>

### Key points (1-3 sentences)
  The paper begins by shortly introducing graph theory and its application to sequencing technology, particularly Hamiltonian cycles and their computational shortcomings (time and space). In response, de Bruijn graphs are proposed as an alternative, which assign k-mers to edges rather than nodes (representing Euler cycles). The paper closes by restating the usefulness of Euler cycles/de Bruijn graphs, but mentions that the propositions require some assumptions (and therefore are not necessarily the best solution to a sequencing problem).

### Added to reading (1-2 sentences)
  Since we discussed these concepts in class, I had some background knowledge in the reading. However, assuming no background knowledge, I think the article may benefit from a more robust introduction to Hamiltonian and Euler cycles.

### Question(s)
  Towards the end of the paper, some propositions are made about de Bruijn graphs, specifically that analogs of them are useful in various bioinformatics problems (even beyond sequencing). In these problems, what are the situational changes that are done to de Bruijn graphs that make them more suitable? 
