<center>
  <h1>Reading Response | Sections 8.1 & 8.2</h1>
</center>

### Key points (1-3 sentences)
These sections discussed evolutionary models and measures of evolutionary distance (8.1) and single phylogenetic tree generation methods (8.2), both of which serve as part of the foundation for reconstructing phylogenetic trees from alignment data. The distance measures discussed range from relatively simple (p-distance) to relatively complex (extensions of p-distance and d_K2P, as part of Kimura two-parameter model), along with their practicality in use. The phylogenetic tree generation methods discussed (such as clustering methods and Fitch-Margoliash method) involve a heavy aura of "no free lunch", with quick points about the advantages and disadvantages of each.

### Added to reading (1-2 sentences)
The reading involved lots of mathematical rigor (in my opinion), and I got lost in it sometimes. I think it would be excellent to include a chart that compares the algorithms visually (advantages and disadvantages of each) towards the end of the corresponding sections.

### Question(s)
In the beginning of section 8.1, it is mentioned that the p-value distance metric for determining evolutionary distance is typically inaccurate, and thus modifications are often necessary for useful results (such as Poisson distance). I have heard of each of these distance metrics in other classes, and I'm curious what other extensions of p-value may be in use today in bioinformatics.
